### ISSUE TITLE
Which sync backend do we use?

### ISSUE THREAD

**Issue by mostlygeek** [mozilla-services/hoverpad#15]
Our choices: 

1. kinto.js w/ a new kinto server cluster
2. webextension chrome.sync backend 
3. sync1.5 

The first requirement is to sync desktop+android firefox clients. iOS would require custom native client code so we will ignore it for now.

**Comment by mostlygeek** [comment]
I think sync 1.5 makes the most sense because: 

* it's already available on FF and Android 
* it is the most mature and stable api we have today
* identity and authentication already done with FxA

**Comment by Natim** [comment]
We don't have any ways to access Sync 1.5 from Webextensions. So I wouldn't consider that being an option.

To answer your question I would say that it depends what you consider Hoverpad to be.
Is it an add-ons for Firefox?

* If the answer is: **yes and only for Firefox users**

I would go with the ``chrome.storage.sync`` API and work on implementing this API for Android.

* If the answer is: **Hoverpad is a tool to sync some HTML/text/image data encrypted between multiple devices**

I would go with Kinto.js

* if we choose to use FxA to authenticate users, the stack is an implementation details but we could use the same kinto-webextensions stack because we can.

* If we want people to be able to sign in with any OAuth providers Github/Facebook/Google/Twitter, we should go with a proper stack.

**Comment by n1k0** [comment]
I'd vote for the storage.sync API for the sake of dogfooding the huge work done by @glasserc on the ffx implementation, but as he expressed concerns with this option he might have very good arguments :)

The kinto.js alternative definitely makes sense if we plan on further integration with the Web and other platforms.

I don't know enough about the Sync api apart from people randomly yelling at it here and there, but I've heard others doing the same with the kinto.js one so that's probably a tie :)

**Comment by ncalexan** [comment]
So: I don't have a complete understanding of what Hoverpad is trying to achieve, but I saw:
1. available in all Firefox's you're signed into;
2. client side encryption.

That means you need to solve essentially the same problem that Firefox Accounts has solved.  That's good!  You can use that work.

In particular, FxA manages two full-strength encryption keys, `kA` and `kB`.  `kA` is truly full-strength and is not  reset when the FxA password is reset; `kB` is full-strength but partially derived from the FxA password, and therefore is reset when the FxA password is reset.  FxA keys are only provided to _verified_ Firefox Accounts (that's the major reason we ask for verification at all.)

If you want these keys, you need to do the FxA sign-in flow documented at https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md.  (Luckily, this is implemented across Desktop, Android, and iOS.)  The reason I emphasize this is that there is another, _much simpler_ way to authenticate FxA-backed services: the oauth flow, documented at https://github.com/mozilla/fxa-oauth-server/blob/master/docs/api.md.  The oauth flow does not, to the best of my knowledge, expose the valuable keys in any way.  (I believe that the oauth flow may have exposed *other* keys to Firefox Hello, but that functionality has now been removed.  Please correct me if I'm wrong.)

So, FxA makes some sense for your problem.  Now, Sync 1.5.  There is a lot of history to consider here, but basically being a "full Sync client" is incredibly delicate.  There are coordination issues (managing `meta/global` and `crypto/keys`) that you just don't want to handle yourself -- or tie into -- if you can avoid them.  There are even authentication issues -- we have a custom token server that adds nothing but implementation overhead!  At the end of the day, Sync 1.5 provides you with a crufty REST API to a shared white-board that's designed for storing tabular data.  You want to expose many clients to a single shared document.  it would be entirely reasonable to only accept edits to the shared document while you have a connection to the service, or to (at least) not work hard to reconcile changes to the document while you're offline.  Sync 1.5 isn't built for this scenario.

Save yourself a lot of pain, forget Sync 1.5 entirely, and use an FxA-oauth authenticated service, and use the FxA managed keys, or derivatives of them.  Perhaps you can get the FxA auth server/oauth server group to help you manage a derivative key set specific to your service, making this type of service significantly easier to spin up.

Finally: you want @rfk and @rnewman on this thread, if only for the history lessons.  Client side encryption has a long and storied history at Mozilla; let's not re-invent past wheels.

**Comment by rnewman** [comment]
@ncalexan: the storage.sync work uses kB, and is already implemented and shipping on desktop.

@mostlygeek: while Sync is implemented on Android and iOS, that doesn't mean your code can access that data in any way. You would need to build a new Sync engine in Swift, and a new Sync engine in Java, and then figure out how to expose that data somehow. You cannot use the Android Sync implementation to sync arbitrary data from a WebExtension.

**Comment by rnewman** [comment]
If you're trying to build a demo for storage.sync, do that, and do the work necessary to make storage.sync work on Android.

If you're actually trying to build an encrypted, shared notepad across platforms… well, I probably wouldn't do that in an add-on. I'd do a native app or a website. There's no need for integration with the browser.

**Comment by rnewman** [comment]
Finally, I'd note that while Sync 1.5 is the longest-lived service like this at Mozilla, the longest-lived people who've worked on it find that it's unsuitable for future purposes, so I wouldn't put too much emphasis on that. It's just not a good platform for the things we want to do these days. I hope and expect that we will have decommissioned Firefox Sync within 5-7 years.

**Comment by leplatrem** [comment]
I arrive late to the party :) It'll be hard not to paraphrase what was already said...

My 2 cents: 

* `storage.sync` would be the best option if we wouldn't want to ship a working demo on desktop+android in a few days/weeks *(gut feeling is that adding its support on Fennec would take months)*
* FxA + `kinto.js` + webcrypto (+WebPush?) would be the pragmatic and universal approach, since it'd work in any Web context. We already have all the pieces as long as we can leverage FxA keys.

Reusing the Webextension server is completely doable but not very clean IMO. A dedicated server would be a no-brainer since it's the exact same setup as the Webextension one. We just mustn't forget to destroy it if the project is killed (in that field we tend to be better at creating than destroying :)).

**Comment by magopian** [comment]
I don't quite understand what the hoverpad/universal notepad is (or is going to be), but here's some context on the current experiment.

Client-side encryption: with the current code on master, we have something that does client-side encryption using webcrypto, on
- standalone web page (works in firefox, chrome, fennec, chrome for android)
- webextension for firefox desktop
- webextension for chrome
- android application: an APK generated through cordova (uses chrome webview)
- linux, osx and windows as standalone applications (using electron)

With the PR #17, which uses older encryption algorithms (md5 of the passphrase for the key instead of a pbkdf2 key derivation, and the AES-CBC algorithm to encrypt/decrypt instead of the newer/stronger AES-GCM algorithm), it also works on
- standalone web page on safari
- ios application: a package generated through cordova (uses webkit)

All this is done using a single codebase.

Currently, it only stores the encrypted data in the localstorage, so we're left with the storage choice. I believe only Kinto would allow us to easily keep this multi-plaform capability. If we don't care about anything other than Firefox and fennec (and we have it implemented in fennec), maybe we could use `chrome.storage.sync`.

I would personally vote for using kinto to keep the ability to run it on every platform/device.

**Comment by Natim** [comment]
Just a note about Kinto and FxA we don't currently have a way to get wrapKB from a FxA Oauth flow which means we cannot really use it. However we could use the current implementation based on a passphrase from the user and a crypto key store in the Kinto in the meantime.

- There are plans to allow grabbing a kA and kB specific to a certain scope using oauth flow.
- There are plans to allow grabbing a FxA bearer token and a kB as par of a web extension API.

**Comment by glasserc** [comment]
Yes, my main concern with using chrome.storage.sync is that it doesn't currently work on Android. The missing pieces are: 

- We need to integrate with Sync somehow, and the Sync implementation on Android is totally different
- We use FxA APIs: `getSignedInUser`, `getOAuthToken`, `sessionStatus`. The FxA implementation on Android is different and may not have exactly these APIs and we would need to add them/expose them to chrome JS.

I'm not sure this is more work than using kinto.js and trying to integrate it with FxA. I do think that integrating kinto.js with a sign-in flow or OAuth flow suits our team's skillset more and can be done with less integration with other teams. On the other hand, I think that getting chrome.storage.sync to work on Android is a laudable goal that would be nice to accomplish while working on this project.

There's one other aspect that I want to bring up here, which is that the chrome.storage.sync API does encryption "transparently", i.e. without any interaction or interference from the extension. I got the impression that we wanted to present the user with a "locked" screen, and decrypt their information once they authenticated somehow. Is that just a UI thing, or is that how we intend to control actual encryption? Did we want to encrypt selected pieces of a notepad or just the whole thing? If we intend to do anything with encryption at all, it seems to me that the advantage of chrome.storage.sync disappears. chrome.storage.sync is IMO going to be a fantastic, bulletproof, cryptographically secure way to store whitelists for your ad blocker, but I don't think it's appropriate for everything.

In short, I think kinto.js provides us the most strategic flexibility, so that's my vote.

**Comment by ncalexan** [comment]
> Yes, my main concern with using chrome.storage.sync is that it doesn't currently work on Android. The missing pieces are:
> 
>     We need to integrate with Sync somehow, and the Sync implementation on Android is totally different

Yep.  There's a cost to building anything Sync-related on Firefox for Android.

>     We use FxA APIs: getSignedInUser, getOAuthToken, sessionStatus. The FxA implementation on Android is different and may not have exactly these APIs and we would need to add them/expose them to chrome JS.

It's totally different, but you could definitely expose the Java equivalents in some way; or we could plumb more of the FxA work into chrome JS and unify the Desktop and Fennec chrome FxA experience.  Both are a lot of work.

> I'm not sure this is more work than using kinto.js and trying to integrate it with FxA. I do think that integrating kinto.js with a sign-in flow or OAuth flow suits our team's skillset more and can be done with less integration with other teams. On the other hand, I think that getting chrome.storage.sync to work on Android is a laudable goal that would be nice to accomplish while working on this project.
> 
> There's one other aspect that I want to bring up here, which is that the chrome.storage.sync API does encryption "transparently", i.e. without any interaction or interference from the extension. I got the impression that we wanted to present the user with a "locked" screen, and decrypt their information once they authenticated somehow. Is that just a UI thing, or is that how we intend to control actual encryption? Did we want to encrypt selected pieces of a notepad or just the whole thing? If we intend to do anything with encryption at all, it seems to me that the advantage of chrome.storage.sync disappears. chrome.storage.sync is IMO going to be a fantastic, bulletproof, cryptographically secure way to store whitelists for your ad blocker, but I don't think it's appropriate for everything.

You could use the cryptographically secure store for "real" security (i.e., full strength keys) and then do some other Crypto Lite :tm: -- PIN based or password based -- for the UI level.  Any attempt to use a PIN or password for "real" security is fraught with peril.

**Comment by grigoryk** [comment]
The not-yet-present `chrome.storage.sync` support on Fennec is an interesting discussion to have and would be a great project to get going, but it should be weighted separately from this discussion, against other Fennec priorities and available resources.

I'm not sure what the product goals are for Hoverpad, but it seems reasonable to strive for "most flexibility" and "least overhead" at an exploratory stage of any product. A standard OAuth flow and some easy-to-use, platform-agnostic storage backend would be my suggestion.

**Comment by rfk** [comment]
> A standard OAuth flow and some easy-to-use, platform-agnostic storage
> backend would be my suggestion.

To work around the issue with getting FxA encryption keys, you could also consider adding FxA *and* a secondary encryption passphrase like you have right now.  It would be an annoying two-step process, but potentially a stepping stone towards a flow where you can use the keys directly from FxA.

**Comment by mostlygeek** [comment]
Hoverpad has two purposes: 

1. Make something appealing enough that people who keep passwords in text files will use it instead. Maybe be a segue for them to a full password manager. 

2. Small, safe and agile way to explore the same challenges we'd face when building a full password manager. 

I like the idea of using kinto.js and FxA OAuth. 

- We don't need to wait for chrome.storage.sync in fennec. 
- We can do more than the key/value store of chrome.storage.sync

Some questions: 

- on desktop/fennec if a user is signed into FxA, can we automatically be authenticated to our backend store? No user interaction required? 
- in fennec, can we reuse the desktop webextension and also use FxA Oauth with no user interaction?
- for standalone mobile apps, can we just integrate FxA auth and load pad?

**Comment by rnewman** [comment]
> on desktop/fennec if a user is signed into FxA, can we automatically be authenticated to our backend store? No user interaction required?

Through a WebExtension, I don't think so. Through the web, no. Through a bootstrapped add-on, sure.

Typically you'd need to go through an OAuth flow of some kind. And I don't think the pieces exist to give you automatic crypto through this avenue yet, but the FxA folks know more about than I do.

> in fennec, can we reuse the desktop webextension and also use FxA Oauth with no user interaction?

I think WebExtensions are on the roadmap, but that Fennec is a little behind desktop. I would be surprised if you could use a desktop WE with no modification — mobile UX is simply different. If a WE for Android is on your radar, you should really talk to Andy McKay and Sebastian to get an accurate picture of where things stand. (And then report back with what you learn! I'd love to hear more.)

> for standalone mobile apps, can we just integrate FxA auth and load pad?

Nothing's ever a 'just'!

The FxA code on both iOS and Android is relatively standalone; making a separate app that logs in with credentials should be straightforward. To use *existing accounts* from Firefox would require some work on Android, and would be difficult or impossible on iOS.

This is the least webby course of action, but the one that is best understood and with the best risk management. You would be committing to two/three shipping vehicles, but every problem you hit would be fixable by you. Mobile development is relatively straightforward.

I get the feeling that Hoverpad is trying to figure out some delivery/crypto/technology questions, not just product questions. It would be useful to know what those constraints are, too.

**Comment by rfk** [comment]
> Typically you'd need to go through an OAuth flow of some kind.

Correct.  If you're already logged in on that device, then you'll probably get a single-click "sign in to hoverpad?" dialog which is not too painful.

> And I don't think the pieces exist to give you automatic crypto through this avenue yet,
> but the FxA folks know more about than I do.

Correct, no crypto keys in the oauth flow yet.  It's something we can work towards but if you want it in Q1 we'd have to have a discussion about prioritization vs other objectives.

**Comment by rnewman** [comment]
> Small, safe and agile way to explore the same challenges we'd face when building a full password manager.

This brings me to a larger point: to make a good decision on software stack, particularly when trail-blazing, it's important to know where you're looking to end up. That is, if you *were* building the full thing, what constraints and objectives would you have?

There are some production questions:

- What would the shipping vehicles be?
- Are you looking to tie this to a browser or not?
- Do you need offline support? Multi-master syncing?
- Do you need to store structured data? Can the schema change? A notepad is very different from Keychain in this regard.
- Should data be recoverable if a user loses their Firefox Account password?
- Do users ever need web access? (E.g., on a borrowed computer.)

For example, if you envision this being part of NMX — a standalone encrypted notes app on your phone, maybe evolving into a password-fill ecosystem participant on Android and iOS — and a desktop browser add-on, then I'd recommend a different approach than if you envision this as a value-add for browsers, intimately tied to web tech. There are other forces if you want web access (a la Dropbox.com).

There are lots of ways you could put together something like what Hoverpad is *now*, which I think the discussion in this issue shows! Not all of those will be useful for the real future goal. So where do you see this ending up?

**Comment by rnewman** [comment]
Prior art that I just stumbled across:

https://standardnotes.org/

**Comment by Natim** [comment]
I believe we will go with Kinto.js for the experiment.
